File: backtracking/39.cpp
/*
 * @Author: httermin
 * @Date: 2019-11-01 16:36:42
 */

#include<string>
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> tmp;
        vector<vector<int>> result;
        getSum(candidates, result, 0, target, tmp);
        return result;
    }

    void getSum(vector<int>& candidates, vector<vector<int>>& result, int start, int target, vector<int>& tmp){
        if(target < 0) return;
        if(target == 0) result.push_back(tmp);
        
        for(int i=start; i<candidates.size(); i++){
            tmp.push_back(candidates[i]);
            getSum(candidates, result, i, target-candidates[i], tmp);
            tmp.pop_back();
        }
    }
};

int main(){
    int a[] = {2, 3, 5, 7};
    vector<int> candidates(a, a+3);

    int n=8;
    Solution ss;

    vector<vector<int>> result = ss.combinationSum(candidates, n);
    for(int i=0; i<result.size(); i++){
        for(int j=0; j<result[i].size(); j++){
            cout<<result[i][j]<<" ";
        }
        cout<<" "<<endl;
    }

    return 1;
}

File: backtracking/39.py
# coding=UTF-8
'''
@Author: httermin
@Date: 2020-01-01 17:38:37
'''
import sys
import os

sys.path.append("..")

from typing import List
from copy import deepcopy

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        tmp = []
        self.getSum(result, candidates, 0, target, tmp)
        return result

    def getSum(self, result, candidates, start, target, tmp):
        if target == 0 and tmp: result.append(tmp[:])
        if target < 0: return 
        
        i = start
        while i < len(candidates):
            
        # for i in range(start, len(candidates)):
            tmp.append(candidates[i])
            if candidates[i] == 0:
                i += 1
            if i < len(candidates):
                self.getSum(result, candidates, i, target-candidates[i], tmp)
            else:
                result.append(tmp[:])
            tmp.pop()
            i += 1

if __name__ == "__main__":
    solu = Solution()
    examples = [([2,3,6,7], 7), ([2,3,5], 8), ([], 1), ([0], 0)]

    for ex, t in examples[:]:
        print(solu.combinationSum(ex, t))